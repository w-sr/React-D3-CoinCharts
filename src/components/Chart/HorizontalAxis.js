import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { extent } from "d3-array";
import { timeFormat } from "d3-time-format";
import styled from "styled-components";

import Flex from "../Flex";
import { DURATION, PROPTYPES } from "../../constants";
import { color, fontSize, fontWeight } from "../../styles/constants";

const StyledHorizontalAxis = styled(Flex).attrs({
  margin: props => (props.hideRightMargin ? "10px 0 0 20px" : "10px 20px 0 20px"),
})`
  margin: ${props => props.margin};
`;

const Tick = styled.div`
  color: ${color.coinchartsGray};
  font-size: ${fontSize.small};
  font-weight: ${fontWeight.medium};
`;

class HorizontalAxis extends PureComponent {
  static formatTime(timestamp, duration) {
    switch (duration) {
      case DURATION.ALL.key:
        return timeFormat("%b %Y")(timestamp); // 'Mmm YYYY'
      case DURATION.YEAR.key:
      case DURATION.MONTH.key:
      case DURATION.WEEK.key:
        return timeFormat("%b %_d")(timestamp); // 'Mmm D'
      case DURATION.DAY.key:
      case DURATION.HOUR.key:
      default:
        return timeFormat("%I:%M %p")(timestamp); // 'HH:MM PM/AM'
    }
  }

  static generateTicks(data, tickCount) {
    if (data.length < 2 || tickCount < 2) {
      return [];
    }

    const [minTime, maxTime] = extent(data, d => d.time);
    const rangeStep = (maxTime - minTime) / (tickCount - 1);
    const generatedTicks = [];
    for (let i = 0; i < tickCount; i += 1) {
      const time = new Date(minTime).valueOf();
      generatedTicks.push(time + i * rangeStep);
    }

    return generatedTicks;
  }

  render() {
    const { data, duration, tickCount, hideRightMargin } = this.props;
    const durationTicks = HorizontalAxis.generateTicks(data, tickCount);
    const axisTicks = durationTicks.map(timestamp => ({
      timestamp,
      label: HorizontalAxis.formatTime(timestamp, duration),
    }));

    return (
      <StyledHorizontalAxis data-testid="HorizontalAxis" justify="space-between" hideRightMargin={hideRightMargin}>
        {axisTicks && axisTicks.map(({ timestamp, label }) => <Tick key={timestamp}>{label}</Tick>)}
      </StyledHorizontalAxis>
    );
  }
}

HorizontalAxis.propTypes = {
  data: PROPTYPES.PRICE_DATA.isRequired,
  duration: PROPTYPES.DURATION.isRequired,
  tickCount: PropTypes.number.isRequired,
  hideRightMargin: PropTypes.bool.isRequired,
};

export default HorizontalAxis;
