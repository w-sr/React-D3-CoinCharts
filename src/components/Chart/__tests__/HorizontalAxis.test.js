import React from "react";
import { render } from "react-testing-library";

import HorizontalAxis from "../HorizontalAxis";

const MAX_DATE = 8000000;
const MIN_DATE = 1000000;
const TEST_PROPS = {
  data: [
    { price: 10, time: new Date(MIN_DATE) },
    { price: 20, time: new Date(2000000) },
    { price: 30, time: new Date(3000000) },
    { price: 40, time: new Date(4000000) },
    { price: 50, time: new Date(5000000) },
    { price: 60, time: new Date(6000000) },
    { price: 70, time: new Date(7000000) },
    { price: 80, time: new Date(MAX_DATE) },
  ],
  duration: "day",
  tickCount: 3,
  hideRightMargin: false,
};

describe("<HorizontalAxis />", () => {
  it("renders without crashing", () => {
    const { container } = render(<HorizontalAxis {...TEST_PROPS} />);
    expect(container.firstChild).not.toBeNull();
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders ticks correctly", () => {
    const { container } = render(<HorizontalAxis {...TEST_PROPS} />);
    expect(container.firstChild.children).toHaveLength(TEST_PROPS.tickCount);
  });

  it("does not render ticks when `TEST_PROPS.data` is []", () => {
    const { container } = render(<HorizontalAxis {...TEST_PROPS} data={[]} />);
    expect(container.firstChild.children).toHaveLength(0);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe("formatTime()", () => {
    const timestamp0 = Date.parse("Sun, 31 Dec 2017 12:00:00");
    const timestamp1 = Date.parse("Mon, 01 Jan 2018 00:00:00");

    it("formats timestamp to `Mmm YYYY` when `duration` = ALL", () => {
      expect(HorizontalAxis.formatTime(timestamp0, "all")).toEqual("Dec 2017");
      expect(HorizontalAxis.formatTime(timestamp1, "all")).toEqual("Jan 2018");
    });

    it("formats timestamp to `Mmm YYYY` when `duration` = YEAR", () => {
      expect(HorizontalAxis.formatTime(timestamp0, "year")).toEqual("Dec 31");
      expect(HorizontalAxis.formatTime(timestamp1, "year")).toEqual("Jan  1");
    });

    it("formats timestamp to `Mmm YYYY` when `duration` = MONTH", () => {
      expect(HorizontalAxis.formatTime(timestamp0, "month")).toEqual("Dec 31");
      expect(HorizontalAxis.formatTime(timestamp1, "month")).toEqual("Jan  1");
    });

    it("formats timestamp to `Mmm YYYY` when `duration` = WEEK", () => {
      expect(HorizontalAxis.formatTime(timestamp0, "week")).toEqual("Dec 31");
      expect(HorizontalAxis.formatTime(timestamp1, "week")).toEqual("Jan  1");
    });

    it("formats timestamp to `Mmm YYYY` when `duration` = DAY", () => {
      expect(HorizontalAxis.formatTime(timestamp0, "day")).toEqual("12:00 PM");
      expect(HorizontalAxis.formatTime(timestamp1, "day")).toEqual("12:00 AM");
    });

    it("formats timestamp to `Mmm YYYY` when `duration` = HOUR", () => {
      expect(HorizontalAxis.formatTime(timestamp0, "hour")).toEqual("12:00 PM");
      expect(HorizontalAxis.formatTime(timestamp1, "hour")).toEqual("12:00 AM");
    });
  });

  describe("generateTicks()", () => {
    const data = [
      { price: 1, time: new Date(1000000) },
      { price: 2, time: new Date(2000000) },
      { price: 2, time: new Date(3000000) },
      { price: 2, time: new Date(5000000) },
      { price: 2, time: new Date(6000000) },
      { price: 2, time: new Date(9000000) },
      { price: 2, time: new Date(10000000) },
    ];

    it("generates a sequence of ticks based on `tickCounts`", () => {
      expect(HorizontalAxis.generateTicks(data, 2)).toEqual([1000000, 10000000]);
      expect(HorizontalAxis.generateTicks(data, 3)).toEqual([1000000, 5500000, 10000000]);
      expect(HorizontalAxis.generateTicks(data, 4)).toEqual([1000000, 4000000, 7000000, 10000000]);
      expect(HorizontalAxis.generateTicks(data, 5)).toEqual([1000000, 3250000, 5500000, 7750000, 10000000]);
    });

    it("returns [] when `tickCount` is invalid (1, 0, negative number)", () => {
      expect(HorizontalAxis.generateTicks(data, 1)).toEqual([]);
      expect(HorizontalAxis.generateTicks(data, 0)).toEqual([]);
      expect(HorizontalAxis.generateTicks(data, -1)).toEqual([]);
    });
  });
});
