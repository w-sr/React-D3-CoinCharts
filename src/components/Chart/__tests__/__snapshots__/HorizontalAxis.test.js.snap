// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HorizontalAxis /> does not render ticks when \`TEST_PROPS.data\` is [] 1`] = `
<div
  class="sc-bwzfXH dPUqxH sc-bdVaJa bJonzb"
  data-testid="HorizontalAxis"
/>
`;

exports[`<HorizontalAxis /> renders without crashing 1`] = `
<div
  class="sc-bwzfXH dPUqxH sc-bdVaJa bJonzb"
  data-testid="HorizontalAxis"
>
  <div
    class="sc-htpNat cghrJv"
  >
    04:16 PM
  </div>
  <div
    class="sc-htpNat cghrJv"
  >
    05:15 PM
  </div>
  <div
    class="sc-htpNat cghrJv"
  >
    06:13 PM
  </div>
</div>
`;
